community/bisect_test.go

```diff
@@ -0,0 +1,271 @@
+	//
+	g := simple.NewUndirectedGraph(0, 0)
+	for u, e := range smallDumbell {
+		for v := range e {
+			g.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v), W: 1})
+		}
+	}
+
+	// Get the profile of internal node weight for resolutions
+	// between 0.1 and 10 using logarithmic bisection.
+	p, err := Profile(ModularScore(g, Weight, 10, nil), true, 1e-3, 0.1, 10)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Modularize for each step midpoint.
+	for _, d := range p {
+		resolution := (d.High + d.Low) / 2
+		r := Modularize(g, resolution, nil)
+		comm := r.Communities()
+		for _, c := range comm {
+			sort.Sort(ordered.ByID(c))
+		}
+		sort.Sort(ordered.BySliceIDs(comm))
+		fmt.Printf("%+.3v: Q=%.3v communities: %v\n", d, Q(r, nil, resolution), comm)
+	}
+
+	// Output:
+	// {Low:0.1 High:0.286 Score:14}: Q=0.807 communities: [[0 1 2 3 4 5]]
+	// {Low:0.286 High:2.33 Score:12}: Q=0.202 communities: [[0 1 2] [3 4 5]]
+	// {Low:2.33 High:3.5 Score:4}: Q=-0.459 communities: [[0 1] [2] [3] [4 5]]
+	// {Low:3.5 High:10 Score:0}: Q=-1.17 communities: [[0] [1] [2] [3] [4] [5]]
+}
+
+var friends, enemies *simple.UndirectedGraph
+
+func init() {
+	friends = simple.NewUndirectedGraph(0, 0)
+	for u, e := range middleEast.friends {
+		// Ensure unconnected nodes are included.
+		if !friends.Has(simple.Node(u)) {
+			friends.AddNode(simple.Node(u))
+		}
+		for v := range e {
+			friends.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v), W: 1})
+		}
+	}
+	enemies = simple.NewUndirectedGraph(0, 0)
+	for u, e := range middleEast.enemies {
+		// Ensure unconnected nodes are included.
+		if !enemies.Has(simple.Node(u)) {
+			enemies.AddNode(simple.Node(u))
+		}
+		for v := range e {
+			enemies.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v), W: -1})
+		}
+	}
+}
+
+func ExampleProfile_multiplex() {
+	// The undirected graphs, friends and enemies, are the political relationships
+	// in the Middle East as described in the Slate article:
+	// http://www.slate.com/blogs/the_world_/2014/07/17/the_middle_east_friendship_chart.html
+	g, err := NewUndirectedLayers(friends, enemies)
+	if err != nil {
+		log.Fatal(err)
+	}
+	weights := []float64{1, -1}
+
+	// Get the profile of internal node weight for resolutions
+	// between 0.1 and 10 using logarithmic bisection.
+	p, err := Profile(ModularMultiplexScore(g, weights, true, WeightMultiplex, 10, nil), true, 1e-3, 0.1, 10)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Modularize for each step midpoint.
+	for _, d := range p {
+		resolution := (d.High + d.Low) / 2
+		r := ModularizeMultiplex(g, weights, []float64{resolution}, true, nil)
+		comm := r.Communities()
+		for _, c := range comm {
+			sort.Sort(ordered.ByID(c))
+		}
+		sort.Sort(ordered.BySliceIDs(comm))
+		fmt.Printf("%+.3v: Q=%.3v communities: %v\n", d, QMultiplex(r, nil, weights, []float64{resolution}), comm)
+	}
+
+	// Output:
+	// {Low:0.1 High:0.717 Score:26}: Q=[20.8 8.04] communities: [[0] [1 7 9 12] [2 8 11] [3 4 5 10] [6]]
```

Comment:
Just as a note of interest, this example shows the randomised nature of the analysis. You can see steps in the profile below that get the same community structure when a single run is performed. This would not happen if we could guarantee the optimal modularisation - the scores given for two identical modularisations by WeightMultiplex should be identical.